### ISSUE TITLE
Undocumented "function overloading" for goto(); and move(); etc. in CBOT

### ISSUE THREAD

**Issue by melex750** [colobot/colobot#673]
It is possible to do "function overloading" with pre-defined functions in CBOT
like `goto();` and `move();` , but the parameter **types** must be different **or** in
a different order.  It works only if defined before `extern void object::`
To call these custom functions from another function, that other function
must have `object::` before its name.

this works:

``` c++

public int goto( int cat ); //            <-- custom goto function.
{
    if (cat == nan) return 1;

    object item = search(cat);

    if (item == null) return 1;
    else return goto(item.position);
}

extern void object::BotMain( ) //         <-- bot's main function.
{

    goto(BotFactory); //              <-- bot goes to nearest BotFactory.

}
```

This should be spelled out in SatCom / programming help / functions / Overloading section.

It is also possible to override pre-defined functions if declared with the same parameters
and you can still call the pre-defined function from that override function.

**Comment by krzys-h** [comment]
This is interesting. And cool. We'll have to play with that and properly document it.

**Comment by MrSimbax** [comment]
Interesting discovery, but does it really needs to be documented? The built-in functions are... functions, so I see no reason why they wouldn't be overloadable, so it's kind of an "obvious" fact.

**Comment by krzys-h** [comment]
The fact that you can call original function should be documented, I:d expect it to recurse into itself.
Does recursion work in CBot anyway?

**Comment by MrSimbax** [comment]
Why would it recurse if there are different parameters for overloaded function and original so they are distinguishable?

**Comment by tomaszkax86** [comment]
Recursion works in CBot. There's even "Stack overflow" error when internal stack overflows, you can see it if you execute this code:

```
extern void object::Test()
{
    int x = factorial(100);

    message(x);
}

int factorial(int n)
{
    if (n < 2)
    {
        return 1;
    }
    else
    {
        return n * factorial(n - 1);
    }
}
```

**Comment by krzys-h** [comment]
@MrSimbax
>It is also possible to override pre-defined functions if declared with the same parameters
and you can still call the pre-defined function from that override function.

**Comment by MrSimbax** [comment]
Override? That doesn't seem to me like a proper behaviour.

**Comment by krzys-h** [comment]
That's why we should document it properly...
I guess it makes a bit of sense if you imagine that the program is inside something like
```Java
public class UserProgram extends ProgramWithStandardFunctions
```

**Comment by MrSimbax** [comment]
If it's not proper, then it should be changed, not documented :P

**Comment by krzys-h** [comment]
Depends if we consider this proper or not. I'd say it's unexpected, but looks correct if explained properly.

**Comment by melex750** [comment]
I see two issues here:
1. the ability to extend a command adding functionality, and maybe error handling
2. the ability to replace a command with your own implementation using `motor(); radar(); space();`

I think the point of documenting it would be to describe how to do it,  and what could be done with it.

Maybe this is intended behavior to allow the user to take the training wheels off,  as a method of learning.

Before this I already wrote a `go();` function with the option to leave the motors on while searching for another point. its faster than goto in most ways, and the program can do things between point a   and  point b  like check resources, detect enemies, and make decisions.


This function looks like it should recurse but it didn't, nomatter what I tried.
even with `object::` no errors.
```c++
public int goto(point position)
{
message("goto override");
return goto(position);
}

extern
```

**Comment by MrSimbax** [comment]
I don't feel like documenting it is a good idea. This seems unintentional and odd. I can mention in the help file that built-in functions can be overloaded, but nothing more. Especially this weird (no) recursion behaviour needs to be further investigated before any attempt to document it.

**Comment by MrSimbax** [comment]
https://github.com/colobot/colobot-data/commit/474ab31b1ab0a56b85179b62524a3caa80641555

**Comment by melex750** [comment]
I found a few days ago, this doesn't work anymore. Evidence in issue #710.

After seeing the source code, I think overloading built-in funcs, was not originally intended.

I am almost certain that replacing CBotString fixed it.

It still works if you disable [this line](https://github.com/colobot/colobot/blob/dev/src/CBot/CBotCStack.cpp#L370), as a test.

If it were to be implemented, comparing parameters in [that function](https://github.com/colobot/colobot/blob/dev/src/CBot/CBotExternalCall.cpp#L69), may be possible.

**Comment by hexagonrecursion** [comment]
> this works:
> 
> ```c++
> public int goto( int cat ); //            <-- custom goto function.
> {
>     if (cat == nan) return 1;
> 
>     object item = search(cat);
> 
>     if (item == null) return 1;
>     else return goto(item.position);
> }
> 
> extern void object::BotMain( ) //         <-- bot's main function.
> {
> 
>     goto(BotFactory); //              <-- bot goes to nearest BotFactory.
> 
> }
> ```

This example now produces a compile error: "Function already exists" (game version 0.2.1-alpha). I do not know when or why this changed.

**Comment by hexagonrecursion** [comment]
Found a typo in your example
```diff
- public int goto( int cat );
+ public int goto( int cat )
```

**Comment by hexagonrecursion** [comment]
This works as originally described in colobot-gold-0.1.6-alpha (yes, even the strange absence of recursion). Starting with colobot-gold-0.1.7-alpha we get a compilation error "Function already exists" instead.

**Comment by hexagonrecursion** [comment]
> This works as originally described in colobot-gold-0.1.6-alpha (yes, even the strange absence of recursion). Starting with colobot-gold-0.1.7-alpha we get a compilation error "Function already exists" instead.

I have bisected this down to this commit: https://github.com/colobot/colobot/commit/9ec61d93e5a432e93f86666628cc084361d13089

**Comment by hexagonrecursion** [comment]
#### TLDR: this was a bug, not a feature. It was fixed. We should close this issue.

I see what happened. There was a bug at line 145
https://github.com/colobot/colobot/blob/87a34ba1ff595812c9248736b7b595dbd57ed7de/src/CBot/CBotCall.cpp#L139-L149

`name == p->GetName()` was converting `CBotString` to `const char*` and then using `operator==(const char*, const char*)` (verified in gdb). The result was always false.

`bool operator==(const CBotString& str)` did not participate in overload resolution because it defined as a member of CBotString rather than a non-member operator overload. Only members of the type of the first operator participate in operator overload resolution. The existence of a converting constructor `CBotString(const char* p)` does not change this.
